Cubical/HITs/Hopf.agda

```diff
@@ -18,16 +38,286 @@ HopfS² (surf i j) = Glue S¹ (λ { (i = i0) → _ , idEquiv S¹
+Hopf→Join (merid y j , x) = filler-1 i1 j y x
+
+-- Backward direction
+Join→Hopf : join S¹ S¹ → TotalSpace
+Join→Hopf (inl x) = (north , x)
+Join→Hopf (inr x) = (south , x)
+Join→Hopf (push y x j) =
+  (merid (inv y * x) j
+  , glue (λ { (j = i0) → y ; (j = i1) → x }) (rotInv-2 x y j))
+
+-- Now for the homotopies, we will need to fill squares indexed by x y : S¹ with value in S¹
+-- Some will be extremeley tough, but happen to be easy when x = y = base
+-- therefore, we fill them for x = y = base and then use the connectedness of S¹ × S¹ and
+-- the discreteness of ΩS¹ to get general fillers.
+
+-- To proceed with that strategy, we first need a lemma :
+-- the sections of the trivial fibration λ (_ : S¹) (_ : S¹) → Int are constant
+
+-- this should be generalized to a constant fibration over a connected space with
+-- discrete fiber
+fibInt : S¹ → S¹ → Set
+fibInt _ _ = Int
+
+-- TODO: this should be moved to a more general place, like HLevel.agda
+mapToSet : (A B : Set) (p : isSet B) → isSet (A → B)
+mapToSet A B p a b x y i j z = p (a z) (b z) (λ i → x i z) (λ i → y i z) i j
+
+S¹→Set : (A : Set) (p : isSet A) (F : S¹ → A) (x : S¹) → F base ≡ F x
+S¹→Set A p F base = refl {x = F base}
+S¹→Set A p F (loop i) = f' i
+  where
+  f : PathP (λ i → F base ≡ F (loop i)) refl (cong F loop)
+  f i = λ j → F (loop (i ∧ j))
+  L : cong F loop ≡ refl
+  L = p (F base) (F base) (f i1) refl
+  f' : PathP (λ i → F base ≡ F (loop i)) (refl {x = F base}) (refl {x = F base})
+  f' = transport (λ i → PathP (λ j → F base ≡ F (loop j)) refl (L i)) f
+
+constant-loop : (F : S¹ → S¹ → Int) → (x y : S¹) → F base base ≡ F x y
+constant-loop F x y = compPath L0 L1
+  where
+  p : isSet (S¹ → Int)
+  p = mapToSet S¹ Int isSetInt
+  L : F base ≡ F x
+  L = S¹→Set (S¹ → Int) p F x
+  L0 : F base base ≡ F x base
+  L0 i = L i base
+  L1 : F x base ≡ F x y
+  L1 = S¹→Set Int isSetInt (F x) y
+
+discretefib : (F : S¹ → S¹ → Set) → Set
+discretefib F = (a : (x y : S¹) → F x y) →
+        (b : (x y : S¹) → F x y) →
+        (a base base ≡ b base base) →
+        (x y : S¹) → a x y ≡ b x y
+
+discretefib-fibInt : discretefib fibInt
+discretefib-fibInt a b h x y i =
+  hcomp (λ t → λ { (i = i0) → constant-loop a x y t
+                 ; (i = i1) → constant-loop b x y t })
+        (h i)
+
+-- first homotopy
+
+assocFiller-3-aux : I → I → I → I → S¹
+assocFiller-3-aux x y j i =
+  hfill (λ t → λ { (i = i0) → rotInv-1 (loop y) (loop (~ y) * loop x) t
+                 ; (i = i1) → rotInv-3 (loop y) (loop x) t
+                 ; (x = i0) (y = i0) → base
+                 ; (x = i0) (y = i1) → base
+                 ; (x = i1) (y = i0) → base
+                 ; (x = i1) (y = i1) → base })
+        (inc ((rotInv-2 (loop x) (loop y) i) * (inv (loop (~ y) * loop x)))) j
+
+-- assocFiller-3-endpoint is used only in the type of the next function, to specify the
+-- second endpoint.
+-- However, I only need the first endpoint, but I cannot specify only one of them as is.
```

Comment:
Just a thought, what happens if you replace the endpoint below with `_` ?

Reply:
Agda cant infer it :(